src/common/libflux/msg_handler.c

```diff
@@ -62,6 +67,114 @@ static void free_msg_handler (flux_msg_handler_t *mh);
 static size_t matchtag_hasher (const void *key);
 static int matchtag_cmp (const void *key1, const void *key2);
 
+
+static void handler_stack_destroy (struct handler_stack *hs)
+{
+    if (hs->stack)
+        zlistx_destroy (&hs->stack);
+    free (hs);
+}
+
+static void handler_stack_destructor (void **item)
+{
+    if (item && *item) {
+        handler_stack_destroy (*item);
+        *item = NULL;
+    }
+}
+
+static struct handler_stack *handler_stack_create (void)
+{
+    struct handler_stack *hs = calloc (1, sizeof (*hs));
+    return hs;
+}
+
+static int handler_stack_add (struct handler_stack *hs,
+                              flux_msg_handler_t *mh)
+{
+    if (!hs->mh) {
+        hs->mh = mh;
+        return 0;
+    }
+    if (!hs->stack) {
+        /*  Create stack, push current entry */
+        if (!(hs->stack = zlistx_new ())
+            || !zlistx_add_start (hs->stack, hs->mh))
+            return -1;
+    }
+    if (!zlistx_add_start (hs->stack, mh))
+        return -1;
```

Comment:
czmq functions do not set errno, so best to explicitly set `errno = ENOMEM` in these failure cases.

also should the function be named `handler_stack_push()` to make its function slightly more clear?